### PR TITLE
Rename Closure_id, etc. introducing (kind of) new notion of "slots"

### PR DESCRIPTION
**Pull request by mshinwell**
After a long discussion we settled on new names for some existing unsatisfactory names:
- `Closure_id` is now `Function_slot`
- `Var_within_closure` is now `Value_slot` (there were objections to using something derived from "environment").

This PR makes these changes (there are bound to be a few places that have been missed but hopefully not too many).  Additionally, matching the above:
- `Select_closure` is now `Project_function_slot`
- `Project_var` is now `Project_value_slot`.

"Slots" are names (or labels, if you like) that refer to parts of sets of closures.  (The term "slot" was already used in some of the Flambda 2 code but not uniformly.)  Function slots tie code IDs together with a particular assignment of captured variables to values (sometimes referred to as a closure environment) and thus in essence describe the semantics of code.  Value slots identify the captured variables.  A particular slot might appear in more than one expression that creates a set of closures.  Slots are never renamed after creation even if a set of closures in which they occur is duplicated (this was not the case for the equivalent notions using the old names in Flambda 1).  Their scope is that of the whole program.

Another way of thinking about "slots" is that they correspond to subsections of `Closure_tag` blocks at runtime.  One such subsection might describe a function (code pointer + arity, etc) or alternatively the value of a captured variable.  Slots are assigned positions ("indexes" or "offsets") within the `Closure_tag` blocks by the `Slot_offsets` module (which has been renamed from `Closure_offsets`).  At present each slot will be assigned the same position in all layouts of `Closure_tag` blocks in which it occurs (corresponding to different sets of closures constructions), which isn't optimal, but will suffice for now.

For review a quick look should suffice plus a more detailed read of the diffs for `Slot_offsets`.  The majority of this PR was done with `sed`.  We will deal with the review of this PR at the retreat.

Based on #589.

### CODE REVIEW

File: middle_end/flambda2/simplify_shared/slot_offsets.ml [hunk 1]
```diff
@@ -203,54 +203,56 @@ let layout_aux j slot (startenv, acc_slots) =
...
+     second field of a Closure_tag block to find out the start of environment.
+     Thus we add a check here to ensure that the slots start with a function
+     slot at offset 0. *)
   match slots with
-  | (0, Closure _) :: _ -> { startenv; slots; empty_env }
+  | (0, Function_slot _) :: _ -> { startenv; slots; empty_env }
   | _ ->
     Misc.fatal_error
-      "Sets of closures must start with a closure slot at offset 0"
+      "Sets of function slots must start with a function slot at offset 0"
```
Comment (Gbury) [inline middle_end/flambda2/simplify_shared/slot_offsets.ml:249]: In this case, I think it's best to keep "set of closures", or use "block with tag closure_tag"

File: middle_end/flambda2/simplify_shared/slot_offsets.ml [hunk 2]
```diff
@@ -369,12 +370,13 @@ module Greedy = struct
...
-      s.id s.first_slot_after_closures s.first_slot_used_by_envvar
-      print_slot_descs s.unallocated_closure_slots print_slot_descs
-      s.unallocated_env_var_slots print_slots s.allocated_slots
+      "@[<v 2>%d:@ @[<v>first_slot_after_function_slots: %d;@ \
+       first_slot_used_by_value_slots: %d;@ unallocated_function_slots: \
+       @[<v>%a@];@ unallocated_value_slots: @[<v>%a@];@ allocated: \
+       @[<v>%a@]@]@]"
+      s.id s.first_slot_after_function_slots s.first_slot_used_by_value_slots
+      print_slot_descs s.unallocated_function_slots print_slot_descs
+      s.unallocated_value_slots print_slots s.allocated_slots
```
Comment (Gbury) [inline middle_end/flambda2/simplify_shared/slot_offsets.ml:379]: This function should be properly formated, and annotated with the attribute prevent ocamlformat from messing with the formatting.

File: middle_end/flambda2/simplify_shared/slot_offsets.ml [hunk 3]
```diff
@@ -428,27 +432,30 @@ module Greedy = struct
...
     | Removed -> env
     | Unassigned -> (
       slot.pos <- Removed;
       match slot.desc with
-      | Closure _ -> Misc.fatal_error "Closure cannot be removed currently"
-      | Env_var v ->
-        let (info : EO.env_var_info) = EO.Dead_var in
-        EO.add_env_var_offset env v info)
+      | Function_slot _ ->
+        Misc.fatal_error "Function_slot cannot be removed currently"
```
Comment (Gbury) [inline middle_end/flambda2/simplify_shared/slot_offsets.ml:451]: Note: we can actually remove function slots now, so this branch could be updated accordingly.

File: middle_end/flambda2/simplify_shared/slot_offsets.ml [hunk 4]
```diff
@@ -428,27 +432,30 @@ module Greedy = struct
...
+      | Value_slot v ->
+        let (info : EO.value_slot_info) = EO.Dead_value_slot in
+        EO.add_value_slot_offset env v info)
     | Assigned _ ->
-      Misc.fatal_error "Cannot remove closure var which is already assigned"
+      Misc.fatal_error "Cannot remove value slot which is already assigned"
```
Comment (Gbury) [inline middle_end/flambda2/simplify_shared/slot_offsets.ml:456]:
```suggestion
      Misc.fatal_error "Cannot remove slot which is already assigned"
```

File: middle_end/flambda2/simplify_shared/slot_offsets.ml [hunk 5]
```diff
@@ -428,27 +432,30 @@ module Greedy = struct
...
 
-  (* Sets of Closures *)
+  (* Sets of Function_slots *)
```
Comment (Gbury) [inline middle_end/flambda2/simplify_shared/slot_offsets.ml:458]:
```suggestion
  (* Blocks with tag closure_tags *)
```

**Review by Gbury — COMMENTED** [review]
I've reviewed the `slot_offsets.ml` file

File: driver/flambda_backend_args.ml [hunk 6]
```diff
@@ -388,7 +388,7 @@ let mk_dflexpect f =
     \     whose basename matches that of the input .ml file (Flambda 2 only)"
 ;;
 
-let mk_dclosure_offsets f =
+let mk_dslot_offsets f =
   "-dclosure-offsets", Arg.Unit f, " Dump closure offsets (Flambda 2 only)"
```
Comment (lthls) [inline driver/flambda_backend_args.ml:392]: I think the name of the option and the actual flag should be kept in sync.

File: middle_end/flambda2/simplify_shared/slot_offsets.ml (hunk 2 above)
Comment (mshinwell) [reply middle_end/flambda2/simplify_shared/slot_offsets.ml:379]: I'm not sure it's worth it in this case, the format string is fairly simple.

File: middle_end/flambda2/simplify_shared/slot_offsets.ml (hunk 3 above)
Comment (mshinwell) [reply middle_end/flambda2/simplify_shared/slot_offsets.ml:451]: I've added a CR

File: middle_end/flambda2/simplify_shared/slot_offsets.ml (hunk 5 above)
Comment (mshinwell) [reply middle_end/flambda2/simplify_shared/slot_offsets.ml:458]: done

File: driver/flambda_backend_args.ml (hunk 6 above)
Comment (mshinwell) [reply driver/flambda_backend_args.ml:392]: Fixed, that was the intention in fact...

File: middle_end/flambda2/from_lambda/closure_conversion.ml [hunk 7]
```diff
@@ -1456,39 +1458,40 @@ let close_let_rec acc env ~function_declarations
...
+          let value_slot =
             VB.create (Variable.create "generated") Name_mode.normal
           in
-          Closure_id.Map.add closure_id closure_var closure_vars_map)
-        generated_closures closure_vars_map
+          Function_slot.Map.add function_slot value_slot value_slots_map)
+        generated_closures value_slots_map
     in
-    let closure_vars =
+    let value_slots =
```
Comment (lthls) [inline middle_end/flambda2/from_lambda/closure_conversion.ml:1476]: Same as above, again.

File: middle_end/flambda2/simplify_shared/exported_offsets.ml [hunk 8]
```diff
@@ -14,16 +14,16 @@
...
-    external cmx can reference elements from another cmx that the current
-    compilation cannot see, all offsets that occur in the current compilation
-    unit should be re-exported. *)
-
-type closure_info =
-  | Dead_id
-  | Id_slot of
+(** Offsets for function and value slots inside sets of closures. They're
+    computed for elements defined in the current compilation unit by
+    [function_slot_offsets], and read from cmx files for external symbols.
```
Comment (lthls) [inline middle_end/flambda2/simplify_shared/exported_offsets.ml:19]:
```suggestion
    [Slot_offsets], and read from cmx files for external symbols.
```

File: middle_end/flambda2/from_lambda/closure_conversion.ml [hunk 9]
```diff
@@ -1402,15 +1403,15 @@ let close_let_rec acc env ~function_declarations
...
-        let closure_id = Function_decl.closure_id decl in
-        ( Closure_id.Map.add closure_id closure_var closure_vars_map,
-          Closure_id.Map.add closure_id ident ident_map ))
-      (Closure_id.Map.empty, Closure_id.Map.empty)
+        let value_slot = VB.create (Env.find_var env ident) Name_mode.normal in
+        let function_slot = Function_decl.function_slot decl in
+        ( Function_slot.Map.add function_slot value_slot value_slots_map,
+          Function_slot.Map.add function_slot ident ident_map ))
+      (Function_slot.Map.empty, Function_slot.Map.empty)
       function_declarations
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:1415]: Fixed

File: middle_end/flambda2/from_lambda/closure_conversion.ml [hunk 10]
```diff
@@ -1456,39 +1458,40 @@ let close_let_rec acc env ~function_declarations
...
            (Function_declarations.funs
               (Set_of_closures.function_decls set_of_closures)))
-        (Closure_id.Map.keys closure_vars_map)
+        (Function_slot.Map.keys value_slots_map)
     in
-    let closure_vars_map =
-      Closure_id.Set.fold
-        (fun closure_id closure_vars_map ->
-          let closure_var =
+    let value_slots_map =
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:1467]: Fixed

File: middle_end/flambda2/from_lambda/closure_conversion.ml (hunk 7 above)
Comment (mshinwell) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:1476]: Fixed

File: middle_end/flambda2/tests/mlexamples/closure_var_a.ml [hunk 11]
```diff
@@ -1,14 +1,13 @@
-(* Comes with closure_var_b.ml and closure_var_c.ml.
+(* Comes with value_slot_b.ml and value_slot_c.ml.
```
Comment (lthls) [inline middle_end/flambda2/tests/mlexamples/closure_var_a.ml:1]: This change should be reverted, or the file names adapted.

File: middle_end/flambda2/simplify_shared/exported_offsets.ml (hunk 8 above)
Comment (mshinwell) [reply middle_end/flambda2/simplify_shared/exported_offsets.ml]: Fixed

**Review by lthls — APPROVED** [review]

**Pull request merged**